source/server/async_operation.h

```diff
@@ -0,0 +1,74 @@
+#ifndef NIDEVICE_GRPC_DEVICE_ASYNC_OPERATION_H
+#define NIDEVICE_GRPC_DEVICE_ASYNC_OPERATION_H
+
+#include <grpcpp/alarm.h>
+#include <grpcpp/grpcpp.h>
+
+#include <deque>
+#include <memory>
+#include <mutex>
+#include <queue>
+
+namespace nidevice_grpc {
+
+template <typename TResponse, typename TJoinable>
+class ServerWriterReactor : public grpc::ServerWriteReactor<TResponse> {
+  using LockGuard = std::lock_guard<std::recursive_mutex>;
+
+ public:
+  virtual ~ServerWriterReactor()
+  {
+  }
+
+  void OnDone() override
+  {
+    delete this;
+  }
+
+  void OnWriteDone(bool ok) override
+  {
+    LockGuard lock(mutex_);
+    write_ready_ = true;
+    pending_send_.pop();
+    if (ok) {
+      next_write();
+    }
+  }
+
+  void OnCancel() override
+  {
+    LockGuard lock(mutex_);
+    done_ = true;
+    this->Finish(::grpc::Status::OK);
+  }
+
+  void queue_write(const TResponse& response)
+  {
+    LockGuard lock(mutex_);
+    pending_send_.push(response);
+    next_write();
+  }
+
+ private:
+  void next_write()
+  {
+    LockGuard lock(mutex_);
+
+    if (!done_ && write_ready_ && !pending_send_.empty()) {
+      write_ready_ = false;
+      auto& response = pending_send_.front();
+      this->StartWrite(&response);
+    }
+  }
+
+  std::recursive_mutex mutex_;
+  // This class relies on: std::deque does not invalidate on push/pop.
+  std::queue<TResponse, std::deque<TResponse>> pending_send_;
+  bool write_ready_{true};
+  bool done_{false};
+
+ public:
```

Comment:
TODO: make this private with a setter.

Document that the position of this is important. It's last in the member list because it must be destroyed *first*. Otherwise the producer may still try to `queue_write` after `mutex_` etc have been destroyed.